datanode/ratelimit/ratelimit.go

```diff
@@ -0,0 +1,196 @@
+	lmt := tollbooth.NewLimiter(cfg.Rate, &limitOpts)
+	lmt.SetBurst(cfg.Burst)
+
+	// The naughty step limiter could have a different rate/burst but it seemed likely to add
+	// more confusion than it's worth to the configuration & these should be sensible.
+	ns := newNaughtyStep(log, cfg.Rate, cfg.Burst, cfg.BanFor.Duration, cfg.TTL.Duration)
+
+	r := &RateLimit{
+		lmt:         lmt,
+		naughtyStep: ns,
+	}
+	r.cfg.Store(cfg)
+	return r
+}
+
+func (r *RateLimit) ReloadConfig(cfg *Config) {
+	r.log.Info("updating rate limit configuration",
+		logging.String("old", fmt.Sprintf("%v", r.cfg.Load())),
+		logging.String("new", fmt.Sprintf("%v", cfg)))
+
+	r.cfg.Store(cfg)
+	r.lmt.SetBurst(cfg.Burst).
+		SetMax(cfg.Rate)
+	r.naughtyStep.lmt.SetBurst(cfg.Burst).
+		SetMax(cfg.Rate)
+	r.naughtyStep.banFor = cfg.BanFor.Duration
+}
+
+func (r *RateLimit) HTTPMiddleware(next http.Handler) http.Handler {
+	middle := func(w http.ResponseWriter, req *http.Request) {
+		if !r.cfg.Load().Enabled {
+			next.ServeHTTP(w, req)
+			return
+		}
+
+		ip := r.ipForRequest(req)
+
+		if r.naughtyStep.isBanned(ip) {
+			r.expressDisappointment(w, banMsg, http.StatusForbidden)
+			return
+		}
+
+		if httpError := tollbooth.LimitByRequest(r.lmt, w, req); httpError != nil {
+			r.naughtyStep.smackBottom(ip)
+			r.expressDisappointment(w, limitMsg, http.StatusTooManyRequests)
+			return
+		}
+
+		next.ServeHTTP(w, req)
+	}
+	return http.HandlerFunc(middle)
+}
+
+func (r *RateLimit) expressDisappointment(w http.ResponseWriter, msg string, status int) {
+	w.Header().Add("Content-Type", "application/json")
+	w.WriteHeader(status)
+	w.Write([]byte(msg))
+}
+
+func (r *RateLimit) ipForRequest(req *http.Request) string {
+	ip := libstring.RemoteIP(r.lmt.GetIPLookups(), r.lmt.GetForwardedForIndexFromBehind(), req)
+	return libstring.CanonicalizeIP(ip)
+}
+
+func (r *RateLimit) GRPCInterceptor(
+	ctx context.Context,
+	req interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (resp interface{}, err error) {
+	if !r.cfg.Load().Enabled {
+		return handler(ctx, req)
+	}
+
+	// Check if the client gave the secret in the metadata, if so skip rate limiting
+	md, ok := metadata.FromIncomingContext(ctx)
+	if ok {
+		mdSecrets := md.Get("X-Rate-Limit-Secret")
+		for _, mdSecret := range mdSecrets {
+			if mdSecret == secret {
+				return handler(ctx, req)
+			}
+		}
+	}
+
+	// Fish out IP address from context
+	addr, ok := contextutil.RemoteIPAddrFromContext(ctx)
+	if !ok {
+		// If we don't have an IP we can't rate limit
+		return handler(ctx, req)
+	}
+
+	ip, _, err := net.SplitHostPort(addr)
+	if err != nil {
+		ip = addr
+	}
+	ip = libstring.CanonicalizeIP(ip)
+
+	// Check the naughty step
+	if r.naughtyStep.isBanned(ip) {
+		return nil, status.Error(codes.Unavailable, banMsg)
+	}
+
+	setRateLimitXResponseHeaders(ctx, r.log, r.lmt, ip)
+	if r.lmt.LimitReached(ip) {
+		r.naughtyStep.smackBottom(ip)
```

Comment:
isn't smacking banned these days?

Reply:
I guess datanode could ban itself if it smacks too often :)